src/libraries/System.Text.Json/ref/System.Text.Json.cs

```diff
@@ -320,6 +320,8 @@ public sealed partial class JsonSerializerOptions
         public System.Text.Json.JsonNamingPolicy? PropertyNamingPolicy { get { throw null; } set { } }
         public System.Text.Json.JsonCommentHandling ReadCommentHandling { get { throw null; } set { } }
         public System.Text.Json.Serialization.ReferenceHandler? ReferenceHandler { get { throw null; } set { } }
+        [System.Runtime.Versioning.RequiresPreviewFeatures("Polymorphic serialization is currently in preview.")]
```

Comment:
I don't think we should be using requires preview features for this. We intend to ship it on 7. Everything we intend to do in 7 is preview right now. This is not different. Folks already need to opt-in to previews by downloading the previews; there needn't be yet another step.

Reply:
The APIs are guaranteed to be changed by Preview 5 (in particular the `IList<PolymorphicTypeConfiguration>` property in `JsonSerializerOptions` will be replaced in favor of properties in the contract model once #63686 is done). There's nothing about the requires preview features messaging that implies it won't be removed by the time 7 has been shipped. At the same time, there's nothing guaranteeing that the APIs will get approved in API review either. To me it seems like an appropriate tool we could use in this context.

Reply:
It's not what the feature is for. It's for features shipping in RTM that should not be relied on in production / likely to break in subsequent RTM releases. Every feature in a preview release may change; getting feedback on features and reacting to them is one of the key reasons we even have preview releases.

Reply:
(If the feature is going to look entirely different one month later, then just don't merge it for preview 4.)

Reply:
> It's not what the feature is for. It's for features shipping in RTM that should not be relied on in production / likely to break in subsequent RTM releases. Every feature in a preview release may change; getting feedback on features and reacting to them is one of the key reasons we even have preview releases.

Point taken, but FWIW we've done this already, cf. https://github.com/dotnet/runtime/pull/64461. We might want to consider providing guidance on when it is appropriate to use this feature and when it is not.

> (If the feature is going to look entirely different one month later, then just don't merge it for preview 4.)

From my perspective, it is very important that this change does get merged in time for preview 4, primarily because of the infrastructural changes that it brings about. Our intention to ship this as a preview feature was since last week, and no objections were raised. cc @jeffhandley

Reply:
> Our intention to ship this as a preview feature was communicated since last week, and no objections were raised.

Both Tanner and I raised objections last week. Eric also raised objections, though I don't know if to you.

Reply:
> but FWIW we've done this already, cf. https://github.com/dotnet/runtime/pull/64461

I believe we shipped bits for that to some specific parties outside of the preview release channel in which everything is preview implicitly. That's an appropriate use.

Reply:
I don't mind removing the annotations, but I do need to get this merged for Preview 4. Would that be ok with you?

Reply:
Yes. If we think the feature is going to ship in some form in November and this is going to yield meaningful opportunity to experiment and provide feedback, go for it.

Reply:
> primarily because of the infrastructural changes that it brings about

For my edification, why does this require public API? That sounds like an implementation detail that doesn't require new surface area, even if that surface area was the motivation.

Reply:
> For my edification, why does this require public API? That sounds like an implementation detail that doesn't require new surface area, even if that surface area was the motivation.

Just for test coverage really. Certain code paths/reentrancy patterns would never be exercised without the polymorphic APIs. For Preview 5 #63686 is going to layer more infrastructural changes on top of this, so need to make sure nothing gets regressed. One option I considered was marking the APIs internal and testing via reflection, but preview features certainly seemed like the path of least resistance.

Secondarily though, we're also looking to get early feedback from a partner team using the APIs shipped in Preview 4.

Reply:
Removed.

Reply:
_Sorry I was late to this discussion._

I was previously in favor of using the preview attribute here, but @stephentoub's arguments above have convinced me otherwise. And as was stated, COSE had another interesting facet of it being delivered outside of traditional preview channels, so we wanted the extra step there. Even with that, there were objections to applying it on those APIs--and that too might not have been the right call in hindsight.

I'm glad we're calibrating and solidifying our guidelines around when to use `[RequiresPreviewFeatures]` and when not to.